src/piton/featurizers/core.py

```diff
@@ -0,0 +1,420 @@
+                    for this label
+        """
+        pass
+
+    def get_column_name(self, column_idx: int) -> str:
+        """Enable the user to get the name of a column by its index
+
+        Args:
+            column_idx (int): The index of the column
+        """
+        return "no name"
+
+    def is_needs_preprocessing(self) -> bool:
+        """Return TRUE if you must run `preprocess()`. If FALSE, then `preprocess()`
+        should do nothing.
+        """
+        return False
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Return dictionary representation."""
+        return {}
+
+    def from_dict(self, data: Mapping[str, Any]):
+        """Convert dictionary representation of Featurizer to object."""
+        pass
+
+
+FeaturizerType = TypeVar("FeaturizerType", bound=Featurizer)
+
+
+class FeaturizerList:
+    """
+    FeaturizerList consists of a list of Featurizers that will be used to (sequentially)
+    featurize Patients based on their Labels.
+
+    It enables preprocessing of featurizers, featurization, and column name extraction.
+    """
+
+    def __init__(self, featurizers: List[Featurizer]):
+        """Create a `FeaturizerList` from a sequence of featurizers.
+
+        Args:
+            featurizers (List[Featurizer]): The featurizers to use for featurizing patients.
+        """
+        self.featurizers: List[Featurizer] = featurizers
+
+    # TODO - restructure code so that multiprocessing happens within the featurizer
+    # (and thus no need to do a bespoke merge of results here)
+    def preprocess_featurizers(
+        self,
+        database_path: str,
+        labeled_patients: LabeledPatients,
+        num_threads: int = 1,
+    ):
+        """Preprocess `self.featurizers` on the provided set of `labeled_patients`."""
+
+        any_needs_preprocessing: bool = any(
+            featurizer.is_needs_preprocessing()
+            for featurizer in self.featurizers
+        )
+        if not any_needs_preprocessing:
+            return
+
+        patient_ids: List[int] = labeled_patients.get_all_patient_ids()
+        patient_ids_per_thread: List[NDArray[np.int64]] = np.array_split(
+            patient_ids, num_threads
+        )
+        tasks = [
+            (database_path, patient_ids, labeled_patients, self.featurizers)
+            for patient_ids in patient_ids_per_thread
+        ]
+
+        # Preprocess in parallel
+        ctx = multiprocessing.get_context("forkserver")
+        with ctx.Pool(num_threads) as pool:
+            trained_featurizers: List[Featurizer] = [
+                y
+                for x in pool.imap(_run_preprocess_featurizers, tasks)
+                for y in x
+            ]
+        grouped_featurizers: collections.defaultdict = collections.defaultdict(
+            list
+        )
+        for featurizer in trained_featurizers:
+            featurizer_name = featurizer.__class__.__name__
```

Comment:
Why do you need featurizer names here? Why not just use indices from the original list? Featurizer 0, featurizer 1, etc, etc.